### PR TITLE
Lint `transmute::<&[u8], &str>`

### PR DESCRIPTION
**Pull request by sinkuu**

### CODE REVIEW

**Comment by oli-obk** [comment]
I think the suggestion should suggest the unchecked variant, since that doesn't break any code.

**Comment by killercup** [comment]
Yeah, from_utf8 returns a `Result<&str, Utf8Error>` so the suggestion can't be auto-applied. Ideally, we'd suggest `.from_utf()?` in functions that return a Result, but even then we can't be sure the error can be converted.

**Comment by oli-obk** [comment]
I think the gain from this lint is to get rid of an unnecessary `transmute`, and not to fix the user's code which relies on `[u8]` -> `str` being valid in every case. Maybe they already checked for ascii, we don't know.

**Comment by sinkuu** [comment]
I see, will change the suggestion later. I just made it conform with the existing transmute_int_to_char lint which suggests not `from_utf8_unchecked` but `from_utf8`.

**Comment by oli-obk** [comment]
Hmm... good point. Maybe we should change this one *and* the `transmute_int_to_char` to suggest the safe method + an unwrap?

Anyone wanting to skip the safety checks for perf reasons can figure out the `unchecked` variant and the caveats of it.

File: clippy_lints/src/transmute.rs [hunk 1]
```diff
@@ -88,14 +88,32 @@ declare_lint! {
...
     pub TRANSMUTE_INT_TO_CHAR,
     Warn,
     "transmutes from an integer to a `char`"
 }
 
+/// **What it does:** Checks for transmutes from a `&[u8] to a `&str`.
+///
+/// **Why is this bad?** Not every byte slice is a valid UTF-8 string.
+///
+/// **Known problems:** None.
```
Comment (killercup) [inline clippy_lints/src/transmute.rs]: - Speed is a known problem, as `from_utf8` needs to check the slice, and its runtime grows linearly with the slice length. This is also the right point in the documentation to mention `from_utf8_unchecked` as alternative that is as fast as the transmute but has a semantically meaningful name.
- The `unwrap` in the suggestion is also some kind of a problem (assuming the byte slice is user input you should most likely return a `Result::Err`), even though I'd say it's a "good" problem: The program will panic where it before had undefined behavior. This is also something I'd mention in the `TRANSMUTE_INT_TO_CHAR` docs.

File: clippy_lints/src/transmute.rs [hunk 2]
```diff
@@ -88,14 +88,32 @@ declare_lint! {
...
-/// let _: Option<char> = std::char::from_u32(x);
+/// let _ = std::char::from_u32(x).unwrap();
 /// ```
 declare_lint! {
     pub TRANSMUTE_INT_TO_CHAR,
     Warn,
     "transmutes from an integer to a `char`"
 }
 
+/// **What it does:** Checks for transmutes from a `&[u8] to a `&str`.
```
Comment (killercup) [inline clippy_lints/src/transmute.rs]: missing <code>\`</code> after `&[u8]`

**Comment by killercup** [comment]
Thanks for the quick update! Can you add the known problems to the TRANSMUTE_INT_TO_CHAR lint's docs as well?

**Comment by oli-obk** [comment]
Thanks!

**Pull request merged**